grape/tree.py

```diff
@@ -0,0 +1,418 @@
+
+    def walk(self, indent: int=3, level: int=0) -> Iterable[Tuple[str, str]]:
+        '''Generator that walks over the tree.
+
+        Args:
+            indent: The indentation level for the report.
+            level: The recursion depth.
+
+        Returns:
+            prefix: The prefix.
+            value: The value.
+
+        Usage:
+            def print_tree(tree: TreeReportNode):
+                'print the tree'
+                for prefix, value in tree.walk():
+                    print(f'{prefix}{str(value)}')
+        '''
+        prefix : str = self.prefix(indent)
+        yield prefix, self.value
+        for child in self._children:
+            yield from child.walk(indent, level+1)
+
+
+def getopts() -> argparse.Namespace:
+    '''Get the command line options.
+
+    Returns:
+        opts: The command line options.
+    '''
+    argparse._ = args_get_text  # type: ignore
+    base = os.path.basename(sys.argv[0])
+    usage = '\n {0} [OPTIONS]'.format(base)
+    desc = 'DESCRIPTION:{0}'.format('\n  '.join(__doc__.split('\n')))
+    epilog = '''
+EXAMPLES:
+    # ------------------------------------------------
+    # Example 1: Help.
+    # ------------------------------------------------
+        $ {2} {0} -h
+
+    # ------------------------------------------------
+    # Example 2: Tree representation for a grafana server.
+    # ------------------------------------------------
+        $ {2} status -v
+        $ {2} {0} -g 4600
+
+    # ------------------------------------------------
+    # Example 3: With indent level 4.
+    # ------------------------------------------------
+        $ {2} status -v
+        $ {2} {0} -i 4 -g 4600
+
+VERSION:
+   {1}
+'''.format(base, __version__, CLI).strip()
+    afc = argparse.RawTextHelpFormatter
+    parser = argparse.ArgumentParser(formatter_class=afc,
+                                     description=desc[:-2],
+                                     usage=usage,
+                                     epilog=epilog.rstrip() + '\n ')
+    parser.add_argument('-i', '--indent',
+                        action='store',
+                        type=int,
+                        default=3,
+                        help='''\
+Indent level.
+Default is %(default)s.
+ ''')
+    parser.add_argument('--sort',
+                        action='store_true',
+                        help='''\
+Sort the tree data.
+ ''')
+    add_common_args(parser)
+    opts = parser.parse_args()
+    return opts
+
+
+def collect_datasources(root: TreeReportNode, services: dict):
+    '''Collect the datasources nodes.
+
+    Args:
+        root: The root of the display tree.
+        services: The dictionary of services.
+    '''
+    datasources = TreeReportNode('datasources', root)
+    for obj in services['datasources']:
+        name = obj['name']
+        did = obj['id']
+        dtype = obj['type']
+        key = f'{name}:id={did}:type={dtype}'
+        TreeReportNode(key, datasources)
```

Comment:
Since you're explicitly relying on hard coded fields, would it be better of the services were a list of `Service` objects, rather than a dict?

The same applies to the method below.

Reply:
That is a good point but i think that for this application it is overkill. We are simply grabbing the fields defined by grafana from their REST API to create a key.